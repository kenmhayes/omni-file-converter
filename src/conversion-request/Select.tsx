import React, { ReactElement, ChangeEvent } from 'react';
import { Form } from 'react-bootstrap';

/**
 * Required information for a selectable option within the select control
 */
export interface SelectOption {
  id: string,
  displayValue: string,
}

/**
 * Input property args for a select control
 */
export interface SelectProps {
  options: SelectOption[];
  onSelectionChange: (optionId: string) => void;
}

/**
 * Helper function to create react elements for SelectOption
 */
function createOptionElement(option: SelectOption): ReactElement {
  return (
    <option
      value={option.id}
      key={option.id}
    >
      {option.displayValue}
    </option>
  );
}

/**
 * A control for selecting one of multiple options displayed in a drop down list form
 * @returns
 *   a react element
 */
function Select(props: SelectProps) {
  const { onSelectionChange, options } = props;

  const onSelectElementChange = (event: ChangeEvent<HTMLSelectElement>) => {
    const optionId = event.target.value || '';
    onSelectionChange(optionId);
  };

  return (
    <div>
      <Form.Select onChange={onSelectElementChange}>
        <option value="">--Please choose an option--</option>
        {
            options.map((option: SelectOption) => createOptionElement(option))
        }
      </Form.Select>
    </div>
  );
}

export default Select;
