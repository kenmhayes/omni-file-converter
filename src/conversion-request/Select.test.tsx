import React from 'react';
import { create } from 'react-test-renderer';
import Select, { SelectOption } from './Select';
import { fireEvent, render } from '@testing-library/react';

it('renders a select form', () => {
  const options: SelectOption[] = [
    { id: '1', displayValue: 'One' },
    { id: '2' , displayValue: 'Two' }
  ]

  const selectionChangeHandler = jest.fn();

  const component = create(
    <Select options={options} onSelectionChange={selectionChangeHandler} placeholderText='Default'/>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('changes the selected option', () => {
    const options: SelectOption[] = [
        { id: '1', displayValue: 'One' },
        { id: '2' , displayValue: 'Two' }
      ]
    
    const selectionChangeHandler = jest.fn();
    const placeholderText = 'Placeholder';
    const { queryByDisplayValue, getByDisplayValue } = render(
        <Select options={options} onSelectionChange={selectionChangeHandler} placeholderText={placeholderText} />
    );

    expect(queryByDisplayValue(placeholderText)).toBeTruthy();

    // Change to a valid option
    fireEvent.change(getByDisplayValue(placeholderText), { target: { value: '1'}});

    expect(queryByDisplayValue('One')).toBeTruthy();
    expect(selectionChangeHandler).toHaveBeenCalled();
});
