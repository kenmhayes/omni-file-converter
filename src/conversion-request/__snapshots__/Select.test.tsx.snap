// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a select form 1`] = `
<div>
  <select
    className="form-select"
    onChange={[Function]}
  >
    <option
      value=""
    >
      Default
    </option>
    <option
      value="1"
    >
      One
    </option>
    <option
      value="2"
    >
      Two
    </option>
  </select>
</div>
`;
